### PR TITLE
Ocean AKS 2.0 Implementation

### PR DESCRIPTION
**Pull request by pripatra**
_This is the default pull request template. You can customize it by adding a `pull_request_template.md` at the root of your repo or inside the `.github` folder._

# Jira Ticket
https://spotinst.atlassian.net/browse/SPOTAUT-14452

_Include a link to your Jira Ticket_
_Example: [JIRAISS-1234](https://spotinst.atlassian.net/browse/JIRAISS-1234)_

# Demo

_Please add a recording of the feature/bug fix in work. if you added new routes, the recording should show the request and response for each new/changed route_

# Checklist:
- [ ] I have filled relevant self assessment ([NodeJS](https://docs.google.com/forms/d/e/1FAIpQLSfl14u9AOBAmxVJ272tvO7XNuXE-EMvWaGcaRZalu1UAKB7RA/viewform), [Frontend](https://docs.google.com/forms/d/e/1FAIpQLSdiBPNKH81w_EkavihVL8Uwb0j7tP8PwJmLFYm2nCOQxz-1qw/viewform), [Backend](https://docs.google.com/forms/d/e/1FAIpQLSed_PsTJ5-XIWkFL6BSDE2AQRBVPmwc3PAmHZkUn-erzVI37Q/viewform?usp=sf_link))
- [ ] I have run ESlint on my changes and fixed all warnings and errors (**NodeJS & Frontend Services**)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have validated all the requirements in the Jira task were answered
- [ ] I have all neccessary approvals for the design/mini design of this task
- [ ] I have approved the API changes and granular permission patterns (documentation subtask) (**For public services only**)

### CODE REVIEW

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 1]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
             response, self.camel_to_underscore)
 
         return formatted_response["response"]["items"][0]
+
```
Comment (ramrutha497) [inline spotinst_sdk2/clients/ocean/__init__.py]: Divide into regions. Take reference to elastigroup client

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 2]
```diff
@@ -0,0 +1,589 @@
+import json
+from enum import Enum
+from typing import List
+
+none = "d3043820717d74d9a17694c176d39733"
+
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py:6]: Divide into regions. Take reference to ocean aws model

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+class OceanAKS:
+    '''
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: uniformly use """ """ for doc strings

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+            self,
+            max_memory_gib: int = none,
+            max_v_cpu: int = none):
+        self.max_memory_gib=max_memory_gib
+        self.max_v_cpu=max_v_cpu
+
+class Down:
+    """
+    # Arguments
+    max_scale_down_percentage: int
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: mark datatype as number double
(float)

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+    accelerated_networking: AcceleratedNetworking
+    disk_performance: DiskPerformance
+    architectures: List[Architectures]
+    exclude_series: List[str]
+    max_memory_giB: float
+    max_vCpu: int
+    min_memory_giB: float
+    min_vCpu: int
+    min_gpu: int
+    max_gpu: int
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: mark datatype as number double for max and min gpu
(float)

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+            filters: Filters = none):
+        self.filters = filters         
+
+class VirtualNodeGroupTemplate:
+    '''
+    # Arguments
+    name: str
+    ocean_id: str 
+    auto_scale: AutoScale
+    availabilty_zones: List[str]
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change to **availability**_zones

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 7]
```diff
@@ -0,0 +1,589 @@
...
+class LaunchNewNodes:
+    '''
+    # Arguments
+    adjustment: int
+    applicable_vm_sizes: List[str]
+    availability_zones: List[str]
+    min_cores_per_node: int
+    min_memory_giB_per_node: float
+    ocean_id: str
+    preferred_lifecycle: str
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: Add enum class
<html>
<body>


preferredLifecycle | stringEnum: "Spot" "Regular"The preferred life cycle to launch the node. If not specified, the life cycle is chosen accoring to the scaled VNG.
-- | --



</body>
</html>

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 8]
```diff
@@ -0,0 +1,589 @@
...
+
+class Type(Enum):
+    label = "label"
+    annotation = "annotation"
+
+class Operator(Enum):
+    equals = "equals"
+    not_equals = "notEquals"
+    exists = "exists"
+    does_not_exist = "doesNotExist"
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: please check the enum value. 
Given as "notExists" in docs

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 9]
```diff
@@ -0,0 +1,589 @@
...
+
+class Filter:
+    """
+    # Arguments
+    conditions: Conditions
+    scope: Scope
+    """
+    def __init__(
+            self,
+            conditions: Conditions = none):
```
Comment (ramrutha497) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: declare scope: Scope = none

**Review by ramrutha497 — CHANGES_REQUESTED** [review]

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 10]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+            url=self.__base_ocean_vng_url + "/" + vng_id,
+            entity_name="ocean_aks_vng"
+        )
+
+        formatted_response = self.convert_json(
+            response, self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+    
+    def get_all_ocean_vng(self):
```
Comment (ramrutha497) [inline spotinst_sdk2/clients/ocean/__init__.py]: Ocean id path param is missing.

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 11]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+        return self.send_delete(
+            url=self.__base_ocean_vng_url + "/" + vng_id,
+            entity_name="ocean_aks_vng"
+        )
+
+    def import_vng_config(self, node_pool_name: str, ocean_id: str):
+        """
+        Import cluster configuration of an AKS cluster to use in create_ocean_cluster api call 
+
+        # Returns
```
Comment (ramrutha497) [inline spotinst_sdk2/clients/ocean/__init__.py:725]: Define arguments in the doc string

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 12]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+
+        formatted_response = self.convert_json(response, 
+                                               self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]  
+
+    def launch_new_nodes(self, body: azure_ocean.LaunchNewNodes):
+        """
+        Launch new nodes in a Ocean cluster 
+
```
Comment (ramrutha497) [inline spotinst_sdk2/clients/ocean/__init__.py:746]: Define arguments in doc string

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 13]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+        response = self.send_get(
+            url=self.__base_ocean_vng_url + "/vmSizes",
+            entity_name="ocean_aks_vng",
+            query_params=dict(virtualNodeGroupId=vng_id))
+
+        formatted_response = self.convert_json(
+            response, self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+
```
Comment (ramrutha497) [inline spotinst_sdk2/clients/ocean/__init__.py]: Get Cluster Aggregated Detailed Costs
Get Cluster Aggregated Summary Costs

Please implement these APIs. Comment here if there is a reason for not implementing the same

**Review by ramrutha497 — CHANGES_REQUESTED** [review]

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 14]
```diff
@@ -0,0 +1,589 @@
+import json
+from enum import Enum
+from typing import List
+
+none = "d3043820717d74d9a17694c176d39733"
+
+class OceanAKS:
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: Rename this to 'Aks'

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 15]
```diff
@@ -0,0 +1,589 @@
...
+    '''
+    # Arguments
+    headrooms: List[AutoScaleHeadroom]
+    '''
+    def __init__(
+            self,
+            headrooms: List[AutoScaleHeadroom] = none):
+        self.headrooms = headrooms                
+
+class Labels:
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: This doesn't seem correct.

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 16]
```diff
@@ -0,0 +1,589 @@
...
+    windows2019="Windows2019"
+    windows2022="Windows2022"
+
+class NodePoolProperties:
+    '''
+    # Arguments
+    enable_node_public_ip: bool
+    kubernetes_version: str
+    max_pods_per_node: int
+    os_disk_size_gB: int
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: Change parameter to os_disk_size_g_b

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 17]
```diff
@@ -0,0 +1,589 @@
...
+    os_disk_type: OsDiskType
+    os_sKU: OsSKU
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change parameter to os_s_k_u

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 18]
```diff
@@ -0,0 +1,589 @@
...
+        self.max_pods_per_node = max_pods_per_node
+        self.os_disk_size_gB = os_disk_size_gB
+        self.os_disk_type =   os_disk_type
+        self.os_sKU = os_sKU
+        self.os_type = os_type  
+
+class Strategy:
+    '''
+    # Arguments
+    fallback_to_oD: bool
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: parameter case, doesn't seem correct.

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 19]
```diff
@@ -0,0 +1,589 @@
...
+    def __init__(
+            self,
+            key: str = none,
+            value:str = none,
+            effect:Effect = none):
+        self.key = key
+        self.value = value
+        self.effect = effect
+
+class Architectures(Enum):
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: Change to "Architecture"

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 20]
```diff
@@ -0,0 +1,589 @@
...
+
+class AcceleratedNetworking(Enum):
+    enabled="Enabled"
+    disabled="Disabled"
+
+class DiskPerformance(Enum):
+    standard="Standard"
+    premium="Premium"
+
+class VmTypes(Enum):
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change to "VmType"

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 21]
```diff
@@ -0,0 +1,589 @@
...
+    high_performance_compute="highPerformanceCompute" 
+    storage_optimized="storageOptimized"
+    gpu="GPU"                                
+
+class Filters:
+    '''
+    # Arguments
+    accelerated_networking: AcceleratedNetworking
+    disk_performance: DiskPerformance
+    architectures: List[Architectures]
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: As per documentation, it is string, please confirm.

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 22]
```diff
@@ -0,0 +1,589 @@
...
+    exclude_series: List[str]
+    max_memory_giB: float
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change to "max_memory_gi_b"

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 23]
```diff
@@ -0,0 +1,589 @@
...
+    max_vCpu: int
+    min_memory_giB: float
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change to "min_memory_gi_b"

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 24]
```diff
@@ -0,0 +1,589 @@
...
+    min_vCpu: int
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change to min_v_cpu

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 25]
```diff
@@ -0,0 +1,589 @@
...
+    min_gpu: int
+    max_gpu: int
+    min_data: int
+    min_nICs: int
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change to min_n_i_cs

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 26]
```diff
@@ -0,0 +1,589 @@
...
+
+class Filters:
+    '''
+    # Arguments
+    accelerated_networking: AcceleratedNetworking
+    disk_performance: DiskPerformance
+    architectures: List[Architectures]
+    exclude_series: List[str]
+    max_memory_giB: float
+    max_vCpu: int
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: change to max_v_cpu

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 27]
```diff
@@ -0,0 +1,589 @@
...
+            virtual_node_group_template: VirtualNodeGroupTemplate = none):
+        self.aks=aks
+        self.auto_scaler=auto_scaler
+        self.controller_cluster_id=controller_cluster_id
+        self.health=health
+        self.name=name
+        self.scheduling=scheduling
+        self.virtual_node_group_template=virtual_node_group_template
+
+class LaunchNewNodes:
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py:470]: Move this to LaunchNewNodesRequest region

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 28]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+            body=body_json,
+            url=self.__base_ocean_cluster_url,
+            entity_name='ocean_aks')
+
+        formatted_response = self.convert_json(response, 
+                                               self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+    
+    def get_all_ocean_cluster(self):
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: rename to get_all_ocean_clusters()

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 29]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+            url=self.__base_ocean_cluster_url + "/" + ocean_id,
+            entity_name='ocean_aks')
+
+        formatted_response = self.convert_json(
+            response,
+            self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+    
+    def import_cluster_config(self, aks_cluster_name: str, resource_group_name: str):
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: Rename to import_cluster_configuration()

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 30]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+        """
+        Import cluster configuration of an AKS cluster to use in create_ocean_cluster api call 
+
+        # Returns
+        (Object): Ocean API response 
+        """
+
+        response = self.send_post(
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: should use send_post_with_params()

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 10 above)
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: Rename to get_all_ocean_vngs()

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 31]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+        )
+
+        formatted_response = self.convert_json(
+            response, self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+    
+    def get_all_ocean_vng(self):
+        """
+        List the configurations for all Ocean VNGs in the specified account.
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: 1. This should support ocean_id to return VNGs in a specified cluster.
2.  The documentation also need to be corrected.

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 32]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+        (Object): Ocean VNG API response
+        """
+        return self.send_delete(
+            url=self.__base_ocean_vng_url + "/" + vng_id,
+            entity_name="ocean_aks_vng"
+        )
+
+    def import_vng_config(self, node_pool_name: str, ocean_id: str):
+        """
+        Import cluster configuration of an AKS cluster to use in create_ocean_cluster api call
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py:723]: Documentation incorrect.

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 33]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+
+        # Returns
+        (Object): Ocean VNG API response
+        """
+        return self.send_delete(
+            url=self.__base_ocean_vng_url + "/" + vng_id,
+            entity_name="ocean_aks_vng"
+        )
+
+    def import_vng_config(self, node_pool_name: str, ocean_id: str):
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: Rename to import_vng_configuration()

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 34]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+        """
+        Import cluster configuration of an AKS cluster to use in create_ocean_cluster api call 
+
+        # Returns
+        (Object): Ocean API response 
+        """
+
+        response = self.send_post(
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: Should use send_post_with_params call.

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 35]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+            url=self.__base_ocean_vng_url+"/import",
+            entity_name='ocean_aks',
+            query_params=dict(nodePoolName=node_pool_name, oceanId=ocean_id))
+
+        formatted_response = self.convert_json(response, 
+                                               self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]  
+
+    def launch_new_nodes(self, body: azure_ocean.LaunchNewNodes):
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py]: rename "body" parameter to node_config

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 36]
```diff
@@ -0,0 +1,589 @@
...
+    '''
+    # Arguments
+    adjustment: int
+    applicable_vm_sizes: List[str]
+    availability_zones: List[str]
+    min_cores_per_node: int
+    min_memory_giB_per_node: float
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: should be min_memory_gi_b_per_node

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 37]
```diff
@@ -0,0 +1,589 @@
...
+    ocean_id: str
+    preferred_lifecycle: str
+    vngIds: List[str]
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: Variable name should be vng_ids

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 38]
```diff
@@ -457,3 +458,320 @@ def launch_nodes_in_vng(self, ocean_launch_spec_id: str, amount: int):
...
+        """
+        Get allowed vm sizes for a particular VNG
+
+        # Arguments
+        vng_id (String): ID of the Ocean VNG
+
+        # Returns
+        (Object): Array of allowed vm sizes list 
+        """
+        response = self.send_get(
```
Comment (anuragsharma-123) [inline spotinst_sdk2/clients/ocean/__init__.py:780]: Should use send_get_with_params()

File: spotinst_sdk2/models/ocean/azure/__init__.py [hunk 39]
```diff
@@ -0,0 +1,589 @@
...
+            value: str = none):
+        self.type = type
+        self.key = key
+        self.operator = operator
+        self.value = value
+
+class AllMatch:
+    """
+    # Arguments
+    all_matches:  List[Attribute]
```
Comment (anuragsharma-123) [inline spotinst_sdk2/models/ocean/azure/__init__.py]: should be all_match as per the schema

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 1 above)
Comment (pripatra) [reply spotinst_sdk2/clients/ocean/__init__.py]: done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 10 above)
Comment (pripatra) [reply spotinst_sdk2/clients/ocean/__init__.py]: not needed for list vngs call

File: spotinst_sdk2/models/ocean/azure/__init__.py (hunk 21 above)
Comment (pripatra) [reply spotinst_sdk2/models/ocean/azure/__init__.py]: type is enum so have implemented like this

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 38 above)
Comment (pripatra) [reply spotinst_sdk2/clients/ocean/__init__.py:780]: function doesn't exist

**Review by anuragsharma-123 — APPROVED** [review]

**Review by ramrutha497 — APPROVED** [review]

**Review by anuragsharma-123 — APPROVED** [review]

**Pull request merged**